File: ProductOfArr.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach

/*
    We use a temp variable to store the left and right products.
    We fill these values in the resulting array.

*/

public int[] productExceptSelf(int[] nums) {
    int[] result = new int[nums.length];
    for (int i = 0, tmp = 1; i < nums.length; i++) {
        result[i] = tmp;
        tmp *= nums[i];
    }
    for (int i = nums.length - 1, tmp = 1; i >= 0; i--) {
        result[i] *= tmp;
        tmp *= nums[i];
    }
    return result;
}

File: ProductOfArr.py
#Time complexity : O(n)
#Space complexity : O(1)


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        output = [1]*len(nums)
        left = nums[ 0]
        right = nums[-1]
        
        for i in range(1, len(nums)):
            output[i] = left
            left *= nums[i]
        print(output)
        
        
        for i in range(len(nums)-2,-1,-1):
            output[i] *= right
            right *= nums[i]
            
        return output
